oshdb-api/src/main/java/org/heigit/bigspatialdata/oshdb/api/object/OSMContribution.java

```diff
@@ -264,4 +266,25 @@ public int getContributorUserId() {
   public long getChangesetId() {
     return data.changeset;
   }
+
+  @Override
+  public int compareTo(@Nonnull OSMContribution other) {
+    return ComparisonChain.start()
+        .compare(this.getOSHEntity().getType(), other.getOSHEntity().getType())
+        .compare(this.getOSHEntity().getId(), other.getOSHEntity().getId())
+        .compare(this.getTimestamp(), other.getTimestamp())
+        .result();
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    if (!(other instanceof OSMContribution)) {
+      return false;
+    }
+    OSMContribution otherContribution = (OSMContribution) other;
+    return this.getOSHEntity().equals(otherContribution.getOSHEntity())
+        && Objects.equals(this.getEntityBefore(), otherContribution.getEntityAfter())
+        && Objects.equals(this.getEntityAfter(), otherContribution.getEntityBefore())
+        && this.getTimestamp().equals(otherContribution.getTimestamp());
```

Comment:
would checking the timestamp first be faster?

Reply:
also with snapshot?

Reply:
Perhaps a bit, but I thought this way is more logical: is it even a contribution/snapshot of the same OSM object -> is it about the same version -> is it about the same timestamp…

That said, I'm hesitating to merge this `equals` method as is. I've added it because after implementing `compareTo`, the code analysis rightly [complained](https://jenkins.ohsome.org/job/oshdb/job/documentation-groupByEntity/3/spotbugs/new/fileName.764779392/source.112c6af9-d6a5-4079-adcf-30f5540ce071/#272) that the class should also implement `equals` and it to be consistent with `compareTo` (i.e. such that `a.equals(b)` is true if and only if `a.compareTo(b) == 0`). This wasn't the case since the default `equals` method in java compares object references (`a == b`). But after implementing `equals` to match `compareTo`, we're now in a situation where (for example) two snapshots of the same entity at the same timestamp but with differing geometries would be considered _equal_, while they are not. Adding a equality check for the geometry could solve this, but would be super slow because it would have to calculate the objects' geometries first. Similarly, I also have some hard time finding a suitable way to implement `hashCode` (which should be consistent with `equals` and `compareTo` also).

I'm leaning toward removing the above implementation of `equals` while leaving the `compareTo` method as is and adding a javadoc comment to state that it doesn't match the default `equals`, perhaps by adding: _“Note: this class has a natural ordering that is inconsistent with equals."_ :shrug:

Reply:
but can that possibly happen: an object at one timestamp con only have one shape (in our database). It could only happen if you wanted to compare e.g. two geometry creation algorithms but in that case you would explicitly compare the geometry, right. 

The only sideeffect I could think of is that if someone (by error) specifies the same timestamp twice in a snapshot view she would get the same snapshot twice for the 'normal' mapReduce but only once when  `groupByEntity`.

Reply:
Yes it can happen, because the `.getGeometry` returns the _clipped_ geometry of an entity at the specific timestamp. If you use (for example) the `.aggregateByGeometry` feature, then an OSM feature lying on the boundary of two "aggregate-by-polygons" will be returned as two snapshots/contribution objects where only the `.getGeometry` output differs.